tunnel/tunnel.go

```diff
@@ -1175,123 +1173,107 @@ func (c *Client) NewSession(target Target) (_ io.ReadWriteCloser, err error) {
-				default:
-				}
-				return
+		switch reg.Registration.(type) {
+		case *tpb.RegisterOp_Session:
+			session := reg.GetSession()
+			if !session.GetAccept() {
+				return fmt.Errorf("connection %d not accepted by server", session.GetTag())
 			}
-
-			switch reg.Registration.(type) {
-			case *tpb.RegisterOp_Session:
-				session := reg.GetSession()
-				if !session.GetAccept() {
+			tag := session.Tag
+			tID := session.GetTargetId()
+			tType := session.GetTargetType()
+			go func() {
+				if err := c.streamHandler(ctx, tag, Target{ID: tID, Type: tType}); err != nil {
 					select {
-					case errCh <- fmt.Errorf("connection %d not accepted by server", session.GetTag()):
+					case errCh <- err:
 					default:
 					}
-					return
 				}
-				tag := session.Tag
-				tID := session.GetTargetId()
-				tType := session.GetTargetType()
-				go func() {
-					if err := c.streamHandler(ctx, tag, Target{ID: tID, Type: tType}); err != nil {
-						select {
-						case errCh <- err:
-						default:
-						}
-					}
-				}()
-			case *tpb.RegisterOp_Target:
-				target := reg.GetTarget()
-				switch op := target.GetOp(); op {
-				case tpb.Target_ADD:
-					if err := c.addPeerTarget(target); err != nil {
-						errCh <- fmt.Errorf("failed to add peer target: %v", err)
-					}
-				case tpb.Target_REMOVE:
-					if err := c.deletePeerTarget(target); err != nil {
-						errCh <- fmt.Errorf("failed to delete peer target: %v", err)
-					}
-				default:
-					// The rest are acks.
-					if !target.GetAccept() {
-						select {
-						case errCh <- fmt.Errorf("target registration (%s, %s) not accepted by server", target.TargetId, target.TargetType):
-						default:
-						}
-						return
-					}
+			}()
+		case *tpb.RegisterOp_Target:
+			target := reg.GetTarget()
+			switch op := target.GetOp(); op {
+			case tpb.Target_ADD:
+				if err := c.addPeerTarget(target); err != nil {
+					return fmt.Errorf("failed to add peer target: %v", err)
 				}
-			case *tpb.RegisterOp_Subscription:
-				sub := reg.GetSubscription()
-				op := sub.GetOp()
-				if op == tpb.Subscription_SUBCRIBE || op == tpb.Subscription_UNSUBCRIBE {
-					if !sub.GetAccept() {
-						select {
-						case errCh <- fmt.Errorf("subscription request (%s) not accepted by server", sub.TargetType):
-						default:
-						}
-						return
-					}
+			case tpb.Target_REMOVE:
+				if err := c.deletePeerTarget(target); err != nil {
+					return fmt.Errorf("failed to delete peer target: %v", err)
+				}
+			default:
+				if !target.GetAccept() {
+					return fmt.Errorf("target registration (%s, %s) not accepted by server", target.TargetId, target.TargetType)
+				}
+			}
+		case *tpb.RegisterOp_Subscription:
+			sub := reg.GetSubscription()
+			op := sub.GetOp()
+			if op == tpb.Subscription_SUBCRIBE || op == tpb.Subscription_UNSUBCRIBE {
+				if !sub.GetAccept() {
+					return fmt.Errorf("subscription request (%s) not accepted by server", sub.TargetType)
 				}
 			}
 		}
-	}()
+	}
 	return <-errCh
```

Comment:
This code is unreachable because the for loop never breaks, it only returns.

A way you could make an error from a streamHandler cause everything to shut down is to store a context cancel function and error inside the Client struct. In register, before creating the stream, you create a context with a cancel func, you store this cancel func on the Client. The context gets passed c.tc.Register.

```
type Client struct {
    cancelFunc func()
    err error
}

func (c *Client) cancel(err error) {
    c.mu.Lock()
    defer c.mu.Unlock()
    c.cancelFunc()
    if c.err != nil {
        c.err = err
    }
}

func (c *Client) Register(ctx) {
    ...
    ctx, c.cancelFunc = context.WithCancel()
    stream, err := c.tc.Register(ctx, c.cc.Opts...)
    ...
}

func (c *Client) Start(ctx) error {
    ...
    for {
        reg, err := c.rs.Recv()
        if err != nil {
            c.cancel(err)
            return c.err
        }
             ...
             go func () {
                 if err := c.streamHandler(ctx); err != nil {
                     c.cancel(err) // Cancels the stream's context, so the next call on the stream will return an error. This will cause the outer for loop to exit.
                 }
             }()
```

By creating the cancel func at the very beginning you have the ability to cancel the stream wherever you want.

Reply:
Thanks for the suggestion. It looks great in this way. I have updated the code as sugguested.